### ISSUE TITLE
Custom DataGenerator to load diffrent format of image lables

### ISSUE THREAD

**Issue by mwindowshz** [pierluigiferrari/ssd_keras#119]
Hi

I am using data that is saved in diffrent format then xml of VOC
I wrote parse function to read my data
I fill the list of 

I checked that the format of information is saved in the same way as it comes from parse_xml() function
I started running training on PASCAL VOC using ssd300training notebook
when using the original xml training worked as you described
when using my input it did not work meaning, training is very slow and after 120 epoch loss is still high 13.xx
while using data from xml I got to loss of 4.xx 

Training loss reduction rate is very slow, I still did not try running more than 120 epoch maybe it would get lower.

I am trying to understand what data is missing from my parser that causes this difference, since I am running basically the same training on VOC. (later I want to run on my own images, but this is a test scenario to see that the parse function works , for now it does not.)
My function fills 
>   self.filenames
>   self.labels
>   self.classes 
>  self.image_ids
>  self.eval_neutral -here I dont have the values of difficult image or not, so all is set to **False**

I run parse_xml, and my parse function and checked the lists, of labels, filenames etc, and cannot find a difference, it seems to be in the same format.
here is an example of self.filenames, and self.labels after loading 2 images from PASCAL using my method:

**labels:**
![image](https://user-images.githubusercontent.com/17617515/40040283-51a880d0-5822-11e8-9d4c-ce1a64e64bb2.png)
**filenames:**
![image](https://user-images.githubusercontent.com/17617515/40040301-679f567a-5822-11e8-8efd-17b6c6380558.png)

Here are is example of using the parse_xml() function

**lables:**
![image](https://user-images.githubusercontent.com/17617515/40040613-a1470a34-5823-11e8-81c7-d9560c0c411a.png)

**filenames:**
![image](https://user-images.githubusercontent.com/17617515/40040591-8c376da0-5823-11e8-9fa5-fb6b36c8fd49.png)

**As you can see both functions produce the same data in the same format.**

**Can you please point me to the error.**

my function is : 
 ```
   def parse_yolo_labels(self,
                          images_list_file_name=None, VocLabelsDirReplace=True,
                          ret=False,classes=['background',
                           'aeroplane', 'bicycle', 'bird', 'boat',
                           'bottle', 'bus', 'car', 'cat',
                           'chair', 'cow', 'diningtable', 'dog',
                           'horse', 'motorbike', 'person', 'pottedplant',
                           'sheep', 'sofa', 'train', 'tvmonitor']):
        '''
        Arguments:
            images_list_file_name (str): images list file containing full path for each file files are the jpg images
            next to each jpg file there should be a txt file containing the yolo_marks
            Yolo marks are as follows, x,y center of rect, width height of rect
            all the values are relative to image size, this is why we need to read the image to get its dimensions

            VOC labels are not in the same folder as the jpeg but in labels folder, so we can read the corresponding
            label from the correct path - this is the use of next argumen:

           VocLabelsDirReplace (bool):  are VOC files jpg and labels in different folders. if True, then replace the
           'JPEGImages','labels', and load the txt file from labels folder
            ret (bool, optional): Whether or not the image filenames and labels are to be returned.
                Defaults to `False`.

        Returns:
            None by default, optionally the image filenames and labels.
        '''

        # Erase data that might have been parsed before
        self.filenames = []
        self.labels = []
        self.image_ids = []
        self.eval_neutral = []
        self.classes = classes
        with open(images_list_file_name) as f:
            # read one image from list, and add its information
            for line in f:
                # open the txt file, so change the ending of file name to be .txt

                imageFile = str(line.rstrip())
                print(imageFile)
                img = cv2.imread(imageFile)
                height, width, c = img.shape
                # self.filenames.append(imageFile)  only add a file if there are labesl
                current_labels = []
                eval_neutr = []  # We'll store whether a box is annotated as "difficult" here.
                labelFile = imageFile.replace('jpg', 'txt')
                if VocLabelsDirReplace:
                    labelFile = labelFile.replace('JPEGImages', 'labels')

                with open(labelFile) as csvfile:
                    readCSV = csv.reader(csvfile, delimiter=' ')
                    boxes = []
                    for row in readCSV:
                        classID = int(row[0]) + 1  # we add 1 because in ssd 0 is background
                        rectHeight = int(float(row[4]) * height)
                        rectWidth = int(float(row[3]) * width)
                        xmin = int(float(row[1]) * width) - int(rectWidth / 2)
                        ymin = int(float(row[2]) * height) - int(rectHeight / 2)
                        xmax = xmin + rectWidth
                        ymax = ymin + rectHeight
                        imageName = os.path.split(imageFile)
                        
                        box = []
                        
                        item_dict = {'image_name': imageFile,
                                     'class_id': classID,                                     
                                     'xmin': xmin,
                                     'ymin': ymin,
                                     'xmax': xmax,
                                     'ymax': ymax}
                        box = []
                        for item in self.labels_output_format:
                            box.append(item_dict[item])
                        boxes.append(box)

                       
                        current_labels.append(box)
                        eval_neutr.append(False)
                       
                if current_labels.__len__() > 0:  # only add file and labels if there are labels
                    
                    self.labels.append(boxes)
                    self.filenames.append(imageFile)
                    head,tail = os.path.split(imageFile)
                    imgID = os.path.splitext(tail)[0]
                    self.image_ids.append(imgID)
                    self.eval_neutral.append(eval_neutr)

        if ret:  # In case we want to return these
            return self.filenames, self.labels
```

My data for example is
[ssd.zip](https://github.com/pierluigiferrari/ssd_keras/files/2003755/ssd.zip)
full code:
[object_detection_2d_data_generator.zip](https://github.com/pierluigiferrari/ssd_keras/files/2003702/object_detection_2d_data_generator.zip)
(should I send other information? full notebook?)

I am writing to you after trying and training and retraining trying to find the issue.
Tried to find answer in other issues posted but could not find. 
I know that this is a function I wrote, but I hope you can give me a pointer of what is missing.
Hope this is ok.
Thanks in advance,
M.

**Comment by pierluigiferrari** [comment]
It's not clear that your parser is even the problem. Looks like it is producing the lists in the right format. I took a look at the samples from your dataset and those samples mostly contain very small objects. I don't know what anchor box parameter values you used, but maybe they aren't ideal for your dataset. Maybe try to use smaller scaling factors to see if that helps. If it does, that could be a first hint in the right direction. Also, don't only rely on the validation loss to determine the performance of the model, better evaluate the mAP instead.

**Comment by mwindowshz** [comment]
Thanks,
Just to clarify, as I wrote. in order to make sure my parser is not the cause,  I decided to train PASCAL VOC images, with my style of labels. using my parser. if this succeeds then I would check again my data-set.
As you can see from attachments training does not seem to work the same. so what can be the problem? should I increase  learning rate? continue training further more epoches?

**results of training using My parser**
[ssd300_pascal_BOXES_07+12_training_log.txt](https://github.com/pierluigiferrari/ssd_keras/files/2006249/ssd300_pascal_BOXES_07.12_training_log.txt)
**results of training using XML** 
[ssd300_pascal_VOC_XML 07+12_training_log.txt](https://github.com/pierluigiferrari/ssd_keras/files/2006250/ssd300_pascal_VOC_XML.07.12_training_log.txt)


But yes my data set contains small objects.

>  I don't know what anchor box parameter values you used, but maybe they aren't ideal for your dataset. Maybe try to use smaller scaling factors to see if that helps

Is this what you mean:
```
scales_pascal = [0.1, 0.2, 0.37, 0.54, 0.71, 0.88, 1.05] # The anchor box scaling factors used in the original SSD300 for the Pascal VOC datasets
scales_coco = [0.07, 0.15, 0.33, 0.51, 0.69, 0.87, 1.05] # The anchor box scaling factors used in the original SSD300 for the MS COCO datasets
```


> Also, don't only rely on the validation loss to determine the performance of the model, better evaluate the mAP instead.


If I understand this is an evaluation done after training. can this be done during training? 


Thanks
M.

**Comment by pierluigiferrari** [comment]
> in order to make sure my parser is not the cause, I decided to train PASCAL VOC images, with my style of labels. using my parser. if this succeeds then I would check again my data-set. As you can see from attachments training does not seem to work the same. so what can be the problem? should I increase learning rate? continue training further more epoches?

Ok I see. What I would do next is compare the `filenames`, `labels` and `image_id` lists for Pascal VOC created by your parser with those created by my XML parser. That is, not just compare individual samples for correctness, but compare the entire list of annotations and print out if there is a deviation of more than one pixel for any coordinates of any of the ground truth boxes. I suspect that the lists are different somewhere, because if they were identical, then the training progress would have to be the same, too.

> If I understand this is an evaluation done after training. can this be done during training?

I mean, "after training" is whatever you define. You can evaluate the model after 120 epochs, or after 40 epochs, or after 1 epoch. The evaluator doesn't know nor care how long the model you give it has been trained for.
